sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,730 @@
+    az_json_writer* json_writer,
+    az_span component_name,
+    az_span property_name,
+    int32_t ack_code,
+    int32_t ack_version,
+    az_span ack_description);
+
+/**
+ * @brief End a property response payload with confirmation status.
+ *
+ * @note This API should be used in conjunction with
+ * az_iot_pnp_client_twin_begin_property_with_status().
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] json_writer The initialized #az_json_writer to append data to.
+ * @param[in] component_name The name of the component to use with this property payload. If this is
+ * for a root or non-component, this can be #AZ_SPAN_EMPTY.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p json_writer must not be `NULL`.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK The JSON payload was suffixed successfully.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_twin_end_property_with_status(
+    az_iot_pnp_client const* client,
+    az_json_writer* json_writer,
+    az_span component_name);
+
+/**
+ * @brief Read the IoT Plug and Play twin properties version for a specified component.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] json_reader The #az_json_reader to parse through.
+ * @param[in] response_type The #az_iot_pnp_client_twin_response_type representing the message type
+ * associated with the payload.
+ * @param[out] out_version The numeric version of the properties in the JSON payload.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p json_reader must not be `NULL`.
+ * @pre \p out_version must not be `NULL`. It must point to a valid int32_t variable.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK If the function returned a valid version.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_twin_get_property_version(
+    az_iot_pnp_client const* client,
+    az_json_reader* json_reader,
+    az_iot_pnp_client_twin_response_type response_type,
+    int32_t* out_version);
+
+/**
+ * @brief Iteratively Read the IoT Plug and Play twin component properties.
+ *
+ * Note that between calls, the #az_span pointed to by \p out_component_name shall not be modified,
+ * only checked and compared. Internally, the #az_span is only changed if the component name changes
+ * in the JSON document and is not necessarily set every invocation of the function.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] json_reader The #az_json_reader to parse through.
+ * @param[in] response_type The #az_iot_pnp_client_twin_response_type representing the message type
+ * associated with the payload.
+ * @param[out] out_component_name The #az_span* representing the value of the component.
+ * @param[out] out_property_name The #az_json_token* representing the name of the property.
+ * @param[out] out_property_value The #az_json_reader* representing the value of the property.
```

Comment:
I don't think this is a clear description. I will take a look at the intended implementation/design and come back to this, potentially with a suggestion.